### PR TITLE
[Feature] Support VOC Dataset in YOLOv5

### PR DESCRIPTION
**Pull request by Nioolek**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
Need to use mmengine==0.2.0.

Add yolov5 voc training.

official yolov5s voc dataset map result:
image

our yolov5s voc dataset training result:
image

The results show that we can reproduce the mAP results on voc dataset.
More details:
1. Now, we can not use coco metric while test on voc dataset in mmyolo. So, to reproduce the map result,  I convert coco test dataset to voc format.  
2. The rate of mosaic while training on voc dataset less than 1.0. While do not use mosaic, the input should use letterbox to transform img. And the hyperparams of leterbox between mosaic and no mosaic are different.  We add `RandomTransform` to solve this. 
3. Using coco metric on voc dataset will be supported in later version.

`RandomChoice` config:
```
randchoice_mosaic_pipeline = dict(
    type='RandomChoice',
    transforms=[with_mosiac_pipeline, without_mosaic_pipeline],
    prob=[0.85834, 0.14166])
```

## Modification

Reference:
https://wandb.ai/glenn-jocher/YOLOv5_VOC_official/runs/jmfrmlsl?workspace=user-
https://github.com/ultralytics/YOLOv5

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
4. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
5. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDetection or MMClassification.
6. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmyolo/datasets/yolov5_voc.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from typing import Any, Optional
+
+from mmdet.datasets import VOCDataset
+
+from ..registry import DATASETS, TASK_UTILS
+
+
+@DATASETS.register_module()
+class YOLOv5VOCDataset(VOCDataset):
```
Comment (hhaAndroid) [inline mmyolo/datasets/yolov5_voc.py:10]: `class YOLOv5VOCDataset(BatchShapePolicyDataset, VOCDataset):`

File: configs/yolov5/README.md [hunk 2]
```diff
@@ -26,6 +26,21 @@ YOLOv5 is a family of object detection architectures and models pretrained on th
 5. The performance is unstable and may fluctuate by about 0.4 mAP.
 6. `balloon` means that this is a demo configuration.
 
+### VOC
+
+| Backbone | size | SyncBN | AMP | Mem (GB) | box AP(COCO metric) |                                                         Config                                                         |         Download         |
+| :------: | :--: | :----: | :-: | :------: | :-----------------: | :--------------------------------------------------------------------------------------------------------------------: | :----------------------: |
+| YOLOv5-n | 512  |  Yes   | Yes |   1.5    |          x          | [config](https://github.com/open-mmlab/mmyolo/tree/master/configs/yolov5/voc/yolov5_n-v61_syncbn_fast_8xb8-50e_voc.py) | [model](<>) \| [log](<>) |
```
Comment (hhaAndroid) [inline configs/yolov5/README.md:33]: We should keep the training method consistent with the official.

File: configs/yolov5/voc/yolov5_s-v61_syncbn_fast_8xb8-50e_voc.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        loss_bbox=dict(loss_weight=0.02),
+        loss_obj=dict(loss_weight=0.51728, class_weight=0.67198),
+        prior_match_thr=3.3744,
+        obj_level_weights=[4., 1., 0.4]),
+    test_cfg=dict(nms=dict(iou_threshold=0.6)))
+
+albu_train_transforms = _base_.albu_train_transforms
+pre_transform = _base_.pre_transform
+
+mosaic_affine_pipeline = [
```
Comment (hhaAndroid) [inline configs/yolov5/voc/yolov5_s-v61_syncbn_fast_8xb8-50e_voc.py:42]: mosaic_affine_pipeline -> with_mosiac_pipeline

File: configs/yolov5/voc/yolov5_s-v61_syncbn_fast_8xb8-50e_voc.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+                max_rotate_degree=0.0,
+                max_translate_ratio=0.04591,
+                max_shear_degree=0.0,
+                scaling_ratio_range=(1 - affine_scale, 1 + affine_scale),
+                border=(-img_scale[0] // 2, -img_scale[1] // 2),
+                border_val=(114, 114, 114)),
+        ])
+]
+
+no_mosaic_affine_pipeline = [
```
Comment (hhaAndroid) [inline configs/yolov5/voc/yolov5_s-v61_syncbn_fast_8xb8-50e_voc.py:78]: no_mosaic_affine_pipeline -> without_mosaic_pipeline

File: configs/yolov5/voc/yolov5_s-v61_syncbn_fast_8xb8-50e_voc.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        border=(0, 0),
+        border_val=(114, 114, 114)),
+    dict(
+        type='LetterResize',
+        scale=img_scale,
+        allow_scale_up=True,
+        pad_val=dict(img=114))
+]
+
+mosaic_prob_config = dict(
```
Comment (hhaAndroid) [inline configs/yolov5/voc/yolov5_s-v61_syncbn_fast_8xb8-50e_voc.py:94]: mosaic_prob_config -> randchoice_mosaic_pipeline

File: configs/yolov5/voc/yolov5_s-v61_syncbn_fast_8xb8-50e_voc.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+    type='mmcv.RandomChoice',
+    transforms=[mosaic_affine_pipeline, no_mosaic_affine_pipeline],
+    prob=[0.85834, 0.14166])
+
+# enable mixup
```
Comment (hhaAndroid) [inline configs/yolov5/voc/yolov5_s-v61_syncbn_fast_8xb8-50e_voc.py:99]: delete

File: configs/yolov5/voc/yolov5_s-v61_syncbn_fast_8xb8-50e_voc.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+            dict(
+                type=dataset_type,
+                data_root=data_root,
+                ann_file='VOC2012/ImageSets/Main/trainval.txt',
+                data_prefix=dict(sub_data_root='VOC2012/'),
+                filter_cfg=dict(filter_empty_gt=False, min_size=32),
+                pipeline=train_pipeline)
+        ]),
+    collate_fn=dict(type='yolov5_collate'),
+    _delete_=True)
```
Comment (hhaAndroid) [inline configs/yolov5/voc/yolov5_s-v61_syncbn_fast_8xb8-50e_voc.py:150]: put in front

File: configs/yolov5/voc/yolov5_s-v61_syncbn_fast_8xb8-50e_voc.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    dict(
+        type='LoadImageFromFile',
+        file_client_args={{_base_.file_client_args}}),
+    dict(type='YOLOv5KeepRatioResize', scale=img_scale),
+    dict(
+        type='LetterResize',
+        scale=img_scale,
+        allow_scale_up=False,
+        pad_val=dict(img=114)),
+    # need gt_ignore
```
Comment (hhaAndroid) [inline configs/yolov5/voc/yolov5_s-v61_syncbn_fast_8xb8-50e_voc.py:162]: why need gt_ignore? please note

File: configs/yolov5/voc/yolov5_s-v61_syncbn_fast_8xb8-50e_voc.py [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+    param_scheduler=dict(
+        lr_factor=lr_factor,
+        max_epochs=max_epochs,
+        warmup_epochs=3.3835,
+        warmup_momentum=0.59462,
+        warmup_bias_lr=0.18657))
+
+# TODO: support using coco metric in voc dataset
+val_evaluator = dict(
+    type='mmdet.VOCMetric', metric='mAP', eval_mode='11points', _delete_=True)
```
Comment (hhaAndroid) [inline configs/yolov5/voc/yolov5_s-v61_syncbn_fast_8xb8-50e_voc.py:206]: Should the `area` evaluation method be used?

File: mmyolo/datasets/yolov5_coco.py [hunk 10]
```diff
@@ -1,24 +1,17 @@
...
 
 
-@DATASETS.register_module()
-class YOLOv5CocoDataset(CocoDataset):
-    """Dataset for YOLOv5 COCO Dataset.
-
-    We only add `BatchShapePolicy` function compared with CocoDataset. See
-    `mmyolo/datasets/utils.py#BatchShapePolicy` for details
-    """
+class BatchShapeDataset(BaseDetDataset):
```
Comment (hhaAndroid) [inline mmyolo/datasets/yolov5_coco.py:9]: BatchShapePolicyDataset

File: mmyolo/datasets/yolov5_coco.py [hunk 11]
```diff
@@ -57,3 +50,13 @@ def prepare_data(self, idx: int) -> Any:
             return self.pipeline(data_info)
         else:
             return super().prepare_data(idx)
+
+
+@DATASETS.register_module()
+class YOLOv5CocoDataset(BatchShapeDataset, CocoDataset):
```
Comment (hhaAndroid) [inline mmyolo/datasets/yolov5_coco.py:56]: The previous `YOLOv5CocoDataset` should give a deprecation warning

File: configs/yolov5/voc/yolov5_s-v61_syncbn_fast_8xb8-50e_voc.py [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+        border_val=(114, 114, 114)),
+    dict(
+        type='LetterResize',
+        scale=img_scale,
+        allow_scale_up=True,
+        pad_val=dict(img=114))
+]
+
+mosaic_prob_config = dict(
+    type='mmcv.RandomChoice',
```
Comment (hhaAndroid) [inline configs/yolov5/voc/yolov5_s-v61_syncbn_fast_8xb8-50e_voc.py:95]: mmcv.RandomChoice -> RandomChoice

File: configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+        warmup_momentum=0.59462,
+        warmup_bias_lr=0.18657))
+
+custom_hooks = [
+    dict(
+        type='EMAHook',
+        ema_type='ExpMomentumEMA',
+        momentum=0.0001,
+        update_buffers=True,
+        strict_load=False,
```
Comment (hhaAndroid) [inline configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py:208]: please note: why need to modify?

File: mmyolo/datasets/yolov5_coco.py (hunk 11 above)
Comment (Nioolek) [reply mmyolo/datasets/yolov5_coco.py]: > The previous `YOLOv5CocoDataset` should give a deprecation warning

I did not add new `YOLOv5CocoDataset`.  I add file `yolov5_datasets.py`.
The plan is to move `YOLOv5CocoDataset`  to  `yolov5_datasets.py` in the next version.

File: configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+
+# TODO: support using coco metric in voc dataset
+val_evaluator = dict(
+    _delete_=True, type='mmdet.VOCMetric', metric='mAP', eval_mode='area')
+
+test_evaluator = val_evaluator
+
+train_cfg = dict(max_epochs=max_epochs)
+
+load_from = 'https://download.openmmlab.com/mmyolo/v0/yolov5/yolov5_s-v61_syncbn_fast_8xb16-300e_coco/yolov5_s-v61_syncbn_fast_8xb16-300e_coco_20220918_084700-86e02187.pth'  # noqa
```
Comment (hhaAndroid) [inline configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py:220]: Put it in front of the model

File: configs/yolov5/README.md [hunk 15]
```diff
@@ -26,6 +26,21 @@ YOLOv5 is a family of object detection architectures and models pretrained on th
 5. The performance is unstable and may fluctuate by about 0.4 mAP.
 6. `balloon` means that this is a demo configuration.
 
+### VOC
+
+| Backbone | size | SyncBN | AMP | Mem (GB) | box AP(COCO metric) |                                                      Config                                                      |         Download         |
```
Comment (hhaAndroid) [inline configs/yolov5/README.md:31]: delete syncbn column

File: configs/yolov5/voc/yolov5_l-v61_fast_1xb32-50e_voc.py [hunk 16]
```diff
@@ -0,0 +1,25 @@
+_base_ = './yolov5_s-v61_fast_1xb64-50e_voc.py'
+
+deepen_factor = 1.0
+widen_factor = 1.0
+train_batch_size_per_gpu = 32
+train_num_workers = 4
```
Comment (hhaAndroid) [inline configs/yolov5/voc/yolov5_l-v61_fast_1xb32-50e_voc.py:6]: train_num_workers = 8

File: configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+        allow_scale_up=True,
+        pad_val=dict(img=114))
+]
+
+randchoice_mosaic_pipeline = dict(
+    type='RandomChoice',
+    transforms=[with_mosiac_pipeline, without_mosaic_pipeline],
+    prob=[0.85834, 0.14166])
+
+# enable mixup
```
Comment (hhaAndroid) [inline configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py:98]: delete comments

File: configs/yolov5/voc/yolov5_x-v61_fast_1xb32-50e_voc.py [hunk 18]
```diff
@@ -0,0 +1,26 @@
+_base_ = './yolov5_s-v61_fast_1xb64-50e_voc.py'
+
+deepen_factor = 1.33
+widen_factor = 1.25
+train_batch_size_per_gpu = 32
+train_num_workers = 4
```
Comment (hhaAndroid) [inline configs/yolov5/voc/yolov5_x-v61_fast_1xb32-50e_voc.py:6]: train_num_workers = 8

File: configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py (hunk 13 above)
Comment (hhaAndroid) [inline configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py:208]: please note: why need to modify?

File: mmyolo/datasets/yolov5_datasets.py [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+    """Dataset for YOLOv5 COCO Dataset.
+
+    We only add `BatchShapePolicy` function compared with CocoDataset. See
+    `mmyolo/datasets/utils.py#BatchShapePolicy` for details
+    """
+    pass
+
+
+@DATASETS.register_module()
+class YOLOv5VOCDataset(BatchShapePolicyDataset, VOCDataset):
```
Comment (hhaAndroid) [inline mmyolo/datasets/yolov5_datasets.py:66]: Refer to mmdet, add UT

File: configs/yolov5/README.md [hunk 20]
```diff
@@ -26,6 +26,21 @@ YOLOv5 is a family of object detection architectures and models pretrained on th
 5. The performance is unstable and may fluctuate by about 0.4 mAP.
 6. `balloon` means that this is a demo configuration.
 
+### VOC
+
+| Backbone | size | AMP | Mem (GB) | box AP(COCO metric) |                                                      Config                                                      |         Download         |
+| :------: | :--: | :-: | :------: | :-----------------: | :--------------------------------------------------------------------------------------------------------------: | :----------------------: |
+| YOLOv5-n | 512  | Yes |   1.5    |          x          | [config](https://github.com/open-mmlab/mmyolo/tree/master/configs/yolov5/voc/yolov5_n-v61_fast_1xb64-50e_voc.py) | [model](<>) \| [log](<>) |
```
Comment (hhaAndroid) [inline configs/yolov5/README.md:33]: supplement

File: configs/yolov5/README.md [hunk 21]
```diff
@@ -26,6 +26,21 @@ YOLOv5 is a family of object detection architectures and models pretrained on th
...
+| YOLOv5-s | 512  | Yes |   2.7    |          x          | [config](https://github.com/open-mmlab/mmyolo/tree/master/configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py) | [model](<>) \| [log](<>) |
+| YOLOv5-m | 512  | Yes |   5.0    |        70.0         | [config](https://github.com/open-mmlab/mmyolo/tree/master/configs/yolov5/voc/yolov5_m-v61_fast_1xb64-50e_voc.py) | [model](<>) \| [log](<>) |
+| YOLOv5-l | 512  | Yes |   8.1    |          x          | [config](https://github.com/open-mmlab/mmyolo/tree/master/configs/yolov5/voc/yolov5_l-v61_fast_1xb32-50e_voc.py) | [model](<>) \| [log](<>) |
+
+**Note**:
+
+1. Training on VOC dataset need pretrained model which trained on COCO.
```
Comment (hhaAndroid) [inline configs/yolov5/README.md:40]: Add official link description

File: configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py [hunk 22]
```diff
@@ -0,0 +1,220 @@
...
+                                             (206, 182)],
+           [(376, 161), (234, 324), (428, 322)]]
+
+load_from = 'https://download.openmmlab.com/mmyolo/v0/yolov5/yolov5_s-v61_syncbn_fast_8xb16-300e_coco/yolov5_s-v61_syncbn_fast_8xb16-300e_coco_20220918_084700-86e02187.pth'  # noqa
+
+model = dict(
+    bbox_head=dict(
+        head_module=dict(num_classes=20),
+        prior_generator=dict(base_sizes=anchors),
+        loss_cls=dict(loss_weight=0.21638, class_weight=0.5),
```
Comment (hhaAndroid) [inline configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py:34]: The source of these parameters should also be explained

File: configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py [hunk 23]
```diff
@@ -0,0 +1,220 @@
...
+    dict(
+        type='LoadImageFromFile',
+        file_client_args={{_base_.file_client_args}}),
+    dict(type='YOLOv5KeepRatioResize', scale=img_scale),
+    dict(
+        type='LetterResize',
+        scale=img_scale,
+        allow_scale_up=False,
+        pad_val=dict(img=114)),
+    dict(type='LoadAnnotations', with_bbox=True),
```
Comment (hhaAndroid) [inline configs/yolov5/voc/yolov5_s-v61_fast_1xb64-50e_voc.py:162]: dict(type='LoadAnnotations', with_bbox=True, _scope_='mmdet'),

File: configs/yolov5/README.md [hunk 24]
```diff
@@ -26,6 +26,24 @@ YOLOv5 is a family of object detection architectures and models pretrained on th
...
+| YOLOv5-m | 512  | Yes |   12.0   |        70.1         | [config](https://github.com/open-mmlab/mmyolo/tree/master/configs/yolov5/voc/yolov5_m-v61_fast_1xb64-50e_voc.py) | [model](https://download.openmmlab.com/mmyolo/v0/yolov5/yolov5_m-v61_fast_1xb64-50e_voc/yolov5_m-v61_fast_1xb64-50e_voc_20221017_114138-815c143a.pth) \| [log](https://download.openmmlab.com/mmyolo/v0/yolov5/yolov5_m-v61_fast_1xb64-50e_voc/yolov5_m-v61_fast_1xb64-50e_voc_20221017_114138.log.json) |
+| YOLOv5-l | 512  | Yes |   10.0   |        72.9         | [config](https://github.com/open-mmlab/mmyolo/tree/master/configs/yolov5/voc/yolov5_l-v61_fast_1xb32-50e_voc.py) | [model](https://download.openmmlab.com/mmyolo/v0/yolov5/yolov5_l-v61_fast_1xb32-50e_voc/yolov5_l-v61_fast_1xb32-50e_voc_20221017_045500-edc7e0d8.pth) \| [log](https://download.openmmlab.com/mmyolo/v0/yolov5/yolov5_l-v61_fast_1xb32-50e_voc/yolov5_l-v61_fast_1xb32-50e_voc_20221017_045500.log.json) |
+
+**Note**:
```
Comment (hhaAndroid) [inline configs/yolov5/README.md:38]: please make sure  'mmengine>=0.2.0'

File: configs/yolov5/README.md [hunk 25]
```diff
@@ -26,6 +26,24 @@ YOLOv5 is a family of object detection architectures and models pretrained on th
...
+
+1. Training on VOC dataset need pretrained model which trained on COCO.
+2. The performance is unstable and may fluctuate by about 0.4 mAP.
+3. YOLOv5 use COCO metric, while training VOC dataset. We will support to use COCO metric while training VOC dataset.
+4. Need to use 'mmengine>=0.0.2'.
+5. Hyperparameter reference from `https://wandb.ai/glenn-jocher/YOLOv5_VOC_official`.
+6. We use single A100 for training, and the single-GPU batch size is 32 or 64.
```
Comment (hhaAndroid) [inline configs/yolov5/README.md:45]: 6 delete

File: mmyolo/datasets/yolov5_coco.py [hunk 26]
```diff
@@ -57,3 +50,14 @@ def prepare_data(self, idx: int) -> Any:
...
             return super().prepare_data(idx)
+
+
+@DATASETS.register_module()
+class YOLOv5CocoDataset(BatchShapePolicyDataset, CocoDataset):
+    """Dataset for YOLOv5 COCO Dataset.
+
+    We only add `BatchShapePolicy` function compared with CocoDataset. See
+    `mmyolo/datasets/utils.py#BatchShapePolicy` for details
+    TODO: need to merge YOLOv5CocoDataset into yolov5_datasets.py
```
Comment (hhaAndroid) [inline mmyolo/datasets/yolov5_coco.py:61]: delete

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**